src/draw/ElkLayout.js

```diff
@@ -0,0 +1,344 @@
+      .map((node) => ({
+        id: node.raw.id,
+        width: node.raw.drawOption.width,
+        height: node.raw.drawOption.height,
+        x: node.raw.drawOption.x,
+        y: node.raw.drawOption.y,
+        layoutOptions: { ...layoutOptions },
+      }));
+
+    // we get the links that have source and target in the children of `node`
+    // TODO : what about cross-level links ? (ie. from a depth to another)
+    const localLinks = this.__getLinksForChildren(nodes, links, parentNode);
+
+    // convert the edges to the format expected by ELK
+    graph.edges = localLinks.map((link) => ({
+      id: `${link[0]}__${link[1]}`, sources: [link[0]], targets: [link[1]],
+    }));
+
+    // finally calling ELK
+    return ElkLayout.__elk.layout(graph);
+  }
+
+  /**
+   * Write a generated layout at a single level of hierarchy to the graph.
+   * @param {ElkNode} elkNode - previously generated layout with a single level of hierarchy.
+   * @private
+   */
+  __writeSingleDepthLayout(elkNode) {
+    const nodes = new Map(elkNode.children
+      .map((node) => [node.id, {
+        x: node.x, y: node.y,
+      }]));
+
+    this.pluginData.components.forEach((component) => {
+      if (nodes.has(component.id)) {
+        const { x, y } = nodes.get(component.id);
+
+        component.drawOption.x = x;
+        component.drawOption.y = y;
+      }
+    });
+  }
+
+  /**
+   * Compute a node's depth (ie. distance to the root).
+   * @param {NodeData} node - a node.
+   * @returns {number} depth.
+   * @private
+   */
+  __computeNodeDepth(node) {
+    if (!node.parent) {
+      // this is the root
+      return 0;
+    }
+
+    return this.__computeNodeDepth(node.parent) + 1;
+  }
+
+  /**
+   * Keep only the links that are contained within {@link nodes} children.
+   * @param {Map<string,NodeData>} nodes - all the nodes.
+   * @param {ComponentLink[]} allLinks - all the links.
+   * @param {NodeData} parentNode - node whose we want the inner links.
+   * @returns {[string,string][]} inner links represented as pairs of nodes ids.
+   * @private
+   */
+  __getLinksForChildren(nodes, allLinks, parentNode) {
+    // ids of parentNode's children
+    const keptIds = new Set(parentNode.children
+      .map((node) => node.raw?.id));
+    // depth at which we operate
+    const depth = parentNode.depth + 1;
+
+    return allLinks
+    // only links with valid source/target
+      .filter((link) => nodes.get(link.source) !== undefined
+            && nodes.get(link.target) !== undefined)
+      .map((link) => {
+        // we operate only at a given depth.
+        // therefore we climb in the hierarchy if needed
+        const sourceDepth = this.__climbToDepth(nodes.get(link.source), depth);
+        const targetDepth = this.__climbToDepth(nodes.get(link.target), depth);
+
+        return [sourceDepth, targetDepth];
+      })
+    // source and target are different and non-falsy
+      .filter((link) => link[0] !== link[1] && link[0] && link[1])
```

Comment:
remove this

Reply:
done

Reply:
undone; the check was in fact useful. After removing it and running the tests, we have a self-link from `workflow1` to `workflow1`, which is not wanted.

Reply:
Replaced with `.filter((link) => link[0] !== link[1])`

Reply:
(I think I deleted Dimitri comment by mistake, wanting to delete mine)
(It said : Done)

Reply:
Why would we remove this ? It makes the E2E tests crash and without it the demo does not run anymore.